# 习题 4.9

![习题 4.9](https://tva1.sinaimg.cn/large/008i3skNly1graympx4o0j31la09k79c.jpg)

In [1]:
data <- read.table("ex_4_9.txt"); data

X1  X2  Y1  Y2 
1  191 155 179 145
2  195 149 201 152
3  181 148 185 149
4  183 153 188 149
5  176 144 171 142
6  208 157 192 152
7  189 150 190 149
8  197 159 189 152
9  188 152 197 159
10 192 150 187 151
11 179 158 186 148
12 183 147 174 147
13 174 150 185 152
14 190 159 195 157
15 188 151 187 158
16 163 137 161 130
17 195 155 183 158
18 186 153 173 148
19 181 145 182 146
20 175 140 165 137
21 192 154 185 152
22 174 143 178 147
23 176 139 176 143
24 197 167 200 158
25 190 163 187 150

In [2]:
X <- data[,1:2]
Y <- data[,3:4]

1. 协方差阵 $S$：

In [3]:
S <- cov(data); S

X1       X2       Y1        Y2      
X1 95.29333 52.86833  69.66167 46.11167
X2 52.86833 54.36000  51.31167 35.05333
Y1 69.66167 51.31167 100.80667 56.54000
Y2 46.11167 35.05333  56.54000 45.02333

基于 $S$ 做典型相关分析：

In [4]:
cca.s <- cancor(X, Y); cca.s

$cor
[1] 0.7885079 0.0537397

$xcoef
         [,1]        [,2]
X1 0.01154653 -0.02857148
X2 0.01443910  0.03816093

$ycoef
         [,1]        [,2]
Y1 0.01025573 -0.03595605
Y2 0.01637533  0.05349758

$xcenter
    X1     X2 
185.72 151.12 

$ycenter
    Y1     Y2 
183.84 149.24

所以有：

第一对典型变量：

$$
\begin{aligned}
V_1 &= 0.01154653 X_1 + 0.01443910 X_2 \\
W_1 &= 0.01025573 Y_1 + 0.01637533 Y_2
\end{aligned}
$$

第一对典型相关系数 $\rho_1 = 0.788507916294635$.

第二对典型变量：

$$
\begin{aligned}
V_2 &= -0.02857148 X_1 + 0.03816093 X_2 \\
W_2 &= -0.03595605 Y_1 + 0.05349758 Y_2
\end{aligned}
$$

第二对典型相关系数 $\rho_2 = 0.0537397044242775$.

2. 相关系数矩阵 $R$:

In [5]:
R <- cor(data); R

X1        X2        Y1        Y2       
X1 1.0000000 0.7345555 0.7107518 0.7039807
X2 0.7345555 1.0000000 0.6931573 0.7085504
Y1 0.7107518 0.6931573 1.0000000 0.8392519
Y2 0.7039807 0.7085504 0.8392519 1.0000000

In [6]:
data.scale <- scale(data)
data.scale.X <- data.scale[,1:2]
data.scale.Y <- data.scale[,3:4]
data.scale

X1          X2          Y1          Y2         
1   0.54088217  0.52624987 -0.48205960 -0.63189808
2   0.95064139 -0.28753859  1.70912039  0.41132988
3  -0.48351588 -0.42317000  0.11553495 -0.03576782
4  -0.27863627  0.25498705  0.41433222 -0.03576782
5  -0.99571490 -0.96569564 -1.27885232 -1.07899577
6   2.28235884  0.79751269  0.81272858  0.41132988
7   0.33600256 -0.15190718  0.61353040 -0.03576782
8   1.15552099  1.06877551  0.51393131  0.41132988
9   0.23356275  0.11935564  1.31072403  1.45455784
10  0.64332197 -0.15190718  0.31473313  0.26229732
11 -0.68839549  0.93314410  0.21513404 -0.18480038
12 -0.27863627 -0.55880141 -0.98005505 -0.33383295
13 -1.20059451 -0.15190718  0.11553495  0.41132988
14  0.43844236  1.06877551  1.11152585  1.15649271
15  0.23356275 -0.01627577  0.31473313  1.30552527
16 -2.32743236 -1.91511551 -2.27484323 -2.86738656
17  0.95064139  0.52624987 -0.08366324  1.30552527
18  0.02868315  0.25498705 -1.07965414 -0.18480038
19 -0.48351588 -0.83006423 -0.18326233 -0.48286551
20 -1.09815470 -1.50822128 -1.87644687 -1.82415860
21  0.64332197  0.39061846  0.11553495  0.41132988
22 -1.20059451 -1.10132705 -0.58165869 -0.33383295
23 -0.99571490 -1.64385269 -0.78085687 -0.92996321
24  1.15552099  2.15382679  1.60952130  1.30552527
25  0.43844236  1.61130115  0.31473313  0.11326475

基于 $R$ 做典型相关分析：

In [7]:
cca.r <- cancor(data.scale.X, data.scale.Y); cca.r

$cor
[1] 0.7885079 0.0537397

$xcoef
        [,1]       [,2]
X1 0.1127152 -0.2789099
X2 0.1064583  0.2813576

$ycoef
        [,1]       [,2]
Y1 0.1029701 -0.3610078
Y2 0.1098775  0.3589657

$xcenter
           X1            X2 
 1.243450e-16 -6.049328e-16 

$ycenter
           Y1            Y2 
-3.380629e-16 -1.359746e-15

得到：

第一对典型变量：

$$
\begin{aligned}
V_1^* &= 0.1127152 X_1^* + 0.1064583 X_2^* \\
W_1^* &= 0.1029701 Y_1^* + 0.1098775 Y_2^*
\end{aligned}
$$

第一对典型相关系数 $\rho_1 = 0.788507916294635$.

第二对典型变量：

$$
\begin{aligned}
V_2^* &= -0.2789099 X_1^* + 0.2813576 X_2^* \\
W_2^* &= -0.3610078 Y_1^* + 0.3589657 Y_2^*
\end{aligned}
$$

第二对典型相关系数 $\rho_2 = 0.0537397044242769$.

由于样本数据同量纲，所以从协方差阵和相关系数矩阵进行典型相关分析，得到的结果是一样的。

3. 对典型变量进行显著性检验

这里需要自己编写函数实现 （ref: https://blog.csdn.net/Tiaaaaa/article/details/58137522 , https://rstudio-pubs-static.s3.amazonaws.com/553282_c1046a4c0b1a40ac9319f51b6207a9d7.html ）：

In [8]:
corcoef.test <- function(cor, n, p, q) {
    # 相关系数检验
    # Args:
    #   r: 典型相关系数
    #   n: 样本个数 (n > p + q)
    #   p, q: 向量的维数
    # Returns:
    #   显著性检验表格
    
    ev <- cor^2
    ev2 <- 1 - ev
    
    l <- length(ev)
    m <- n - 1 - (p+q+1) / 2
    w <- cbind(NULL)  # 保存中间计算值

    for (i in 1:l) {
      w <- cbind(w,prod(ev2[i:l]))
    }

    Q <- c(NULL); d <- c(NULL)
    for (i in 1:l){
      Q <- cbind(Q, -(m-(i-1)) * log(w[i]))
      d <- cbind(d, (p-i+1) * (q-i+1))
    }

    pvalue <- pchisq(Q, d, lower.tail=FALSE)    # 计算卡方统计量对应的概率
    
    bat <- cbind(t(Q), t(d), t(pvalue))
    colnames(bat) <- c("Chi-Squared", "df", "pvalue")
    rownames(bat) <- 1:l
    
    bat    # ret
}

In [9]:
corcoef.test(cca.r$cor, n=25, p=2, q=2)

Chi-Squared df pvalue      
1 20.96417998 4  0.0003218897
2  0.05928875 1  0.8076236560

取显著水平为 $\alpha=0.05$，其中第一对典型变量的检验 $p$ 值为 $0.003<0.05$，认为第一对典型变量显著相关，
而第二对典型变量的检验 $p$ 值为 $0.8031>0.05$，认为第二对典型变量不是显著相关。